src/sentry/api/event_search.py

```diff
@@ -36,17 +36,31 @@
 
 event_search_grammar = Grammar(
     r"""
-search            = (boolean_operator / paren_term / search_term)*
-boolean_operator  = spaces (or_operator / and_operator) spaces
-paren_term        = spaces open_paren spaces (paren_term / boolean_operator / search_term)+ spaces closed_paren spaces
-search_term       = key_val_term / quoted_raw_search / raw_search
-key_val_term      = spaces (time_filter / rel_time_filter / specific_time_filter
-                    / duration_filter / boolean_filter / numeric_filter
-                    / aggregate_filter / aggregate_date_filter / aggregate_rel_date_filter
-                    / has_filter / is_filter / text_filter)
-                    spaces
-raw_search        = (!key_val_term ~r"\ *(?!(?i)OR(?![^\s]))(?!(?i)AND(?![^\s]))([^\ ^\n ()]+)\ *" )*
-quoted_raw_search = spaces quoted_value spaces
+search = (boolean_operator / paren_term / search_term)*
+
+boolean_operator = spaces (or_operator / and_operator) spaces
+paren_term       = spaces open_paren spaces (paren_term / boolean_operator / search_term)+ spaces closed_paren spaces
+search_term      = key_val_term / free_text_quoted / free_text
+
+free_text        = (spaces !key_val_term !or_operator !and_operator (free_parens / ~r"[^()\n ]+") spaces)+
+free_parens      = spaces open_paren (spaces free_text spaces)? closed_paren spaces
```

Comment:
What does this new term capture?

Reply:
It basically captures a parenthesis expression that _isn't_ a boolean expression, since it contains free text, which can't contain boolean operators or key value terms

Reply:
So if we had a query like `(hello i am free text)` how would this be interpreted before vs after this change?

Reply:
Hmm actually in that case the parens are preserved.

Here's some examples

```py
In [3]: parse_search_query('(thing)')
Out[3]: [ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing'))])]

In [4]: parse_search_query('test(thing)')
Out[4]: [SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test(thing)'))]

In [5]: parse_search_query('test (thing)')
Out[5]: [SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test (thing)'))]

In [6]: parse_search_query('test (thing OR thing2)')
Out[6]:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test')),
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing')), 'OR', SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing2'))])]

In [7]: parse_search_query('test (thing) AND (thing3)')
Out[7]:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test (thing)')),
 'AND',
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing3'))])]

```

Reply:
Ok, looks like this is an improvement on old behaviour:

```
In [5]: parse_search_query('(thing)')
Old:
Out[5]: [ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing'))])]
New:
Out[2]: [ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing'))])]

In [6]: parse_search_query('test(thing)')
Old:
Out[6]:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test')),
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing'))])]
New:
Out[3]: [SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test(thing)'))]

In [7]: parse_search_query('test (thing OR thing2)')
Old:
Out[7]:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test')),
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing')), 'OR', SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing2'))])]
New:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test')),
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing')), 'OR', SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing2'))])]

In [8]: parse_search_query('test (thing) AND (thing3)')
Old:
Out[8]:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test')),
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing'))]),
 'AND',
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing3'))])]
New:
[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='test (thing)')),
 'AND',
 ParenExpression(children=[SearchFilter(key=SearchKey(name='message'), operator='=', value=SearchValue(raw_value='thing3'))])]
```

Reply:
I feel like we probably need some tests in place to cover these changes though? Also still a few BE tests failing

Reply:
Yeah I see that. I did update a number of tests, all are improvements I believe

Reply:
I was thinking more tests in `test_event_search.py` so that the parser is specifically tested. Those other tests go through the transformation from `SearchFilter` to a snuba query, I think

Reply:
Oh you're right, we didn't really have many tests in there for this